infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/server/secure/jwtproxy/JwtProxySecureServerExposer.java

```diff
@@ -68,11 +69,34 @@ public void expose(
       ServicePort servicePort,
       Map<String, ServerConfig> secureServers)
       throws InfrastructureException {
-    ServicePort exposedServicePort =
-        proxyProvisioner.expose(
-            k8sEnv, serviceName, servicePort, servicePort.getProtocol(), secureServers);
 
-    exposer.expose(
-        k8sEnv, machineName, proxyProvisioner.getServiceName(), exposedServicePort, secureServers);
+    AtomicReference<InfrastructureException> exceptionHolder = new AtomicReference<>();
+
+    ExternalServerExposer.onEachExposableServerSet(
+        secureServers,
+        (serverId, servers) -> {
+          if (exceptionHolder.get() != null) {
+            return;
+          }
+
+          try {
+            ServicePort exposedServicePort =
+                proxyProvisioner.expose(
```

Comment:
Why do we need a dedicated port for each exposableServerSet here?
It's kind of surprise me that here we are in a loop of `onEachExposableServerSet` invoke `exposer.expose` that will do iterate on `onEachExposableServerSet` more more time https://github.com/eclipse/che/blob/d46eb20398472a8e2a751002b1ae0ac85b5f56ef/infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/server/external/ExternalServerExposer.java#L97
Did I miss something, if not - can we avoid it?

Reply:
We need a dedicated port, because a server set needs to be exposed uniquely. So we need a unique route so that the user "sees" it as unique, but we also need a unique jwt service (and therefore a port), so that the backing service sees it as a request coming from a unique domain.

Good catch of calling `onEachExposableServerSet` twice unnecessarily. Because of the changes to correctly handle the preview URL, the `exposeAsSingle` method is now public on the server exposer as well, so we can use that instead in the jwt server exposer.

Reply:
> but we also need a unique jwt service (and therefore a port), so that the backing service sees it as a request coming from a unique domain.

Sorry, but I did not get it. Why backing service cares about domain requests come? Could you elaborate more?
In my understanding purpose why we introduce unique servers: for browser not to send all shared cookies...
And I don't understand how using one JwtProxy can break it...